### ISSUE TITLE
Add Secure Sharing of Keys

### ISSUE THREAD

**Issue by smcgregor** [privly/privly-android#35]
Mobile applications are an ideal way to transfer keys from one trusted device to another. In this project you would be building out this functionality for integration with the [Privly family of applications](https://github.com/privly/privly-applications). The ideal proposal would include key exchange with NFC and/or QR codes. The challenge with this project will not be to build communication for key exchange, but to properly integrate the exchanged keys with the rest of the architecture. One potential process would be to:

1. Two users select "exchange contacts" in the Android application
1. The users place their phones near each other and keys are exchanged
1. The exchanged key is signed by the mobile device and pushed to the user's key exchange storage (the content server would work here, but this could equally be Twitter or email)
1. The exchanged key is also stored to the local storage of the webview so privly-applications can benefit from the newly trusted contact.
1. When the user encounters the shared key from the content server (or twitter/email) on the browser version of the extension, the key is synced to the browser.

This is intended to be the discussion point for [project proposals](https://github.com/privly/privly-organization/wiki/Ideas#user-content-advance-android-application-).

**Comment by vshivam** [comment]
So, the first step will be to implement a secure local storage of keys which could be synced with the server and be accessible to local privly applications as well ?

**Comment by smcgregor** [comment]
I think we would want to store the keys in the localstorage context of the
web view. That would make the keys available to all local applications. The
trickey part will be syncing the collection of keys between devices, which
may work with a qr code and websockets. We need to think about how we can
make this usable.

For a student to be successful in this they will need to have an excellent
understanding of how both architectures plug together. We should probably
require a pull request from both platforms.

**Comment by akki12345** [comment]
#smcgregor you say store the keys in the local phone storage context of the web view...this is done.So now what we have to do is.....syncing the collection of keys between devices with a qr code and any type of password(the syncing part is like a Google sync.).

**Comment by arjunvijayvargiya** [comment]
Hi @smcgregor ,
                            Can you please elaborate on the fifth point of the process i.e."When the user encounters the shared key from the content server (or twitter/email) on the browser...."

**Comment by smcgregor** [comment]
The purpose of the last point is to allow for synchronization between the mobile app and the desktop's browser extensions. This could be accomplished in several ways, but the most important aspect is that the keys not be imported to the desktop unless they are trusted by the phone. This will be easy of the desktop and the mobile already share a private key, so it may be better to concentrate on transferring private keys between devices.

**Comment by arjunvijayvargiya** [comment]
we can put the private key on the Google drive /drop box from the same user phone using Google drive API or Drop Box API available for android. Is it possible for the extension to fetch data from Google drive/drop box ,so that it can share the private key with him???

**Comment by arjunvijayvargiya** [comment]
we can put the private key on the Google drive /drop box from the app using Google drive API or Drop Box API available for android. Is it possible for the extension to fetch data from Google drive/drop box ,so that it can share the private key with itself??? Will it be a correct approach???

**Comment by smcgregor** [comment]
Private keys need to remain _private_. If you share the key via Drop box or Google drive, then Drop box and Google Drive also have the private key.

**Comment by arjunvijayvargiya** [comment]
Yeah !!! I forgot to see a major flaw in putting keys on drive/dropbox.

**Comment by arjunvijayvargiya** [comment]
We can create a QR code from the private key using Zxing library available in android.This QR code can be scanned and read by the Online QR code scanner functionality using webcam available on desktop
For example: http://www.webqr.com/index.html 
This tool is a javascript port using the Zxing library. Github Link:https://github.com/LazarSoft/jsqrcode 

Then the extension can take over to fetch the private key

**Comment by arjunvijayvargiya** [comment]
I have one more idea which can be achieved with the help of socket programming. The service will be running with a particular IP (IPv4 or IPv6) and a particular host number. What we can do is when we want to transfer a private key from mobile to desktop extension, we will send it to the defined IP and port number. Since server is in listen mode, it process the information and acknowledges by sending a confirmation which is verified at mobile user side. Only those mobile will be connected which receives a confirmation.Is it a correct approach???

**Comment by arjunvijayvargiya** [comment]
I have one more way can be achieved with the help of socket programming. The service will be running with a particular IP(ipv4 or ipv6) and a particular host number. What we can do is when we want to transfer a key from mobile to desktop extension, we will send it to the defined ip and port number. Since server is in listen mode, it process the information and acknowledges by sending a confirmation which is verified at mobile user side. Only those mobile will be connected which receives a confirmation. Will it be a correct approach.

**Comment by smcgregor** [comment]
One potential approach would be to have a QR code displayed by the desktop that contains a key and a identifier. The QR code key would encrypt mobile's private key in transit and would be destroyed immediately after the private key is decrypted. The identifier would allow both parties to identify their communications channel.

A similar approach could be used for mobile to mobile and desktop to mobile key transfer.

**Comment by arjunvijayvargiya** [comment]
I think this key exchange mechanism would be done when the user registers for a new privly account on the app, as we want the user to have his private key already synced  before exploring on other app functions.

**Comment by mkash32** [comment]
Would it be a good idea to maintain the private key in the privly database so that the private key will be associated with the user? In this manner transferring of private key wouldn't be needed (only for public keys of trusted contacts would need to be transferred). The private key could be symmetrically encrypted with the user's password so it would be secure on the privly server also. 
I was thinking about the situation where a user creates the private key through the mobile device and distributes the public key to his/her friends. If the user happens to delete the app then the private key info is lost. If the user's friends try to send any messages using his public key then it won't be possible to decrypt those messages. Then the user would have to generate another private-public key and redistribute to all of his/her friends.

**Comment by smcgregor** [comment]
Yes, key loss is a huge issue. It could be good to have a backup, it depends on the specifics of the implementation. It is difficult to comment more specifically without a more concrete proposal.
